src/Microsoft.Net.Http.Headers/MediaTypeParser.cs

```diff
@@ -0,0 +1,275 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.Net.Http.Headers
+{
+    public struct MediaTypeParser : IEnumerable<MediaTypeComponent>
+    {
+        private readonly string _mediaType;
+        private readonly int _offset;
+        private readonly int? _length;
+
+        public MediaTypeParser(StringSegment mediaType)
+            : this(mediaType.Buffer, mediaType.Offset, mediaType.Length)
+        {
+        }
+
+        public MediaTypeParser(string mediaType, int offset, int? length)
+        {
+            if (mediaType == null)
+            {
+                throw new ArgumentNullException(nameof(mediaType));
+            }
+
+            if (offset < 0 || offset >= mediaType.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+
+            if (length != null && offset + length > mediaType.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(length));
+            }
+
+            _mediaType = mediaType;
+            _offset = offset;
+            _length = length;
+        }
+
+        public MediaTypeComponent? GetParameter(string parameterName)
+        {
+            return GetParameter(new StringSegment(parameterName));
+        }
+
+        public MediaTypeComponent? GetParameter(StringSegment parameterName)
+        {
+            var componentsEnumerator = GetEnumerator();
+
+            if (!(componentsEnumerator.MoveNext() || componentsEnumerator.MoveNext()))
```

Comment:
The comment is inside the if block, if we fail to enumerate the first two components (type and subtype) we fail to parse the media type